docs/design/diagnostics.md

```diff
@@ -0,0 +1,293 @@
+$ terraform plan
+# ... other plan output ...
+╷
+│ Error: Summary
+│ 
+│   on example.tf line #:
+│    #: source configuration line
+│ 
+│ Details
+╵
+```
+
+This is highly visible feedback to practitioners. Diagnostic severity also influences the exit status of Terraform CLI.
+
+### Terraform Plugin Protocol
+
+The protocol defines the following message types for diagnostics handling:
+
+```protobuf
+message Diagnostic {
+    enum Severity {
+        INVALID = 0;
+        ERROR = 1;
+        WARNING = 2;
+    }
+    Severity severity = 1;
+    string summary = 2;
+    string detail = 3;
+    AttributePath attribute = 4;
+}
+
+message AttributePath {
+    message Step {
+        oneof selector {
+            // Set "attribute_name" to represent looking up an attribute
+            // in the current object value.
+            string attribute_name = 1;
+            // Set "element_key_*" to represent looking up an element in
+            // an indexable collection type.
+            string element_key_string = 2;
+            int64 element_key_int = 3;
+        }
+    }
+    repeated Step steps = 1;
+}
+```
+
+It is found in various RPC `Response` messages, such as:
+
+```protobuf
+message ValidateProviderConfig {
+    message Request {
+        DynamicValue config = 1;
+    }
+    message Response {
+        repeated Diagnostic diagnostics = 2;
+    }
+}
+```
+
+All RPCs except `StopProvider` include diagnostics support.
+
+### terraform-plugin-go
+
+The `terraform-plugin-go` library, which underpins this framework, provides the following implementation (both `tfprotov5` and `tfprotov6`) of diagnostics:
+
+```go
+package tfprotov6
+
+type Diagnostic struct {
+    Severity  DiagnosticSeverity
+    Summary   string
+    Detail    string
+    Attribute *tftypes.AttributePath
+}
+```
+
+These types include no helper methods or slice type alias.
+
+## Prior Implementations
+
+### terraform-plugin-sdk
+
+The previous framework provided a thin abstraction layer in the [`diag` package](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/diag):
+
+```go
+type Diagnostic struct {
+    Severity      Severity
+    Summary       string
+    Detail        string
+    AttributePath cty.Path
+}
+
+func (d Diagnostic) Validate() error
+
+type Diagnostics []Diagnostic
+
+func (diags Diagnostics) HasError() bool
+
+func Errorf(format string, a ...interface{}) Diagnostics
+
+func FromErr(err error) Diagnostics
+```
+
+## Goals
+
+Diagnostics in this framework should be:
+
+- Able to natively support warning and error diagnostics.
+- Implementable by provider developers.
+- Abstracted from terraform-plugin-go and convertable into those types to separate implementation concerns.
+- Ergonomic to create, read, and delete in Go code (e.g. have helper methods for common use cases).
+
+Additional consideration should be given to:
+
+- Extensibility and additional data storage (e.g. distinguishing between types of diagnostics or what generated an error).
```

Comment:
So I think this covers a way to achieve what we want, but I think what we want is to be able to handle diagnostics programmatically, whereas right now they're mostly tailored for a human, practitioner audience.

Also, could you expand on "what generated an error" means here?

Reply:
I like focusing on the word "programmatic" here -- hopefully clarified these with https://github.com/hashicorp/terraform-plugin-framework/pull/108/commits/a29e9cda0d7fd3c5e145103e5986721c7b3eab3d